src/packet/public_key_encrypted_session_key.rs

```diff
@@ -47,61 +62,200 @@ impl PublicKeyEncryptedSessionKey {
+            }
+        }
 
-        let mpis = pkey.encrypt(rng, &data)?;
+        let values = pkey.encrypt(rng, &data, EskType::V3_4)?;
 
-        Ok(PublicKeyEncryptedSessionKey {
+        Ok(PublicKeyEncryptedSessionKey::V3 {
             packet_version: Default::default(),
-            version: 3,
             id: pkey.key_id(),
-            algorithm: pkey.algorithm(),
-            mpis,
+            pk_algo: pkey.algorithm(),
+            values,
+        })
+    }
+
+    /// Encrypts the given session key to the passed in public key, as a v6 pkesk.
+    /// FIXME: cleanup/DRY with from_session_key
+    pub fn from_session_key6<R: CryptoRng + Rng>(
+        rng: &mut R,
+        session_key: &[u8],
+        pkey: &impl PublicKeyTrait,
+    ) -> Result<Self> {
+        // the session key
+        let mut data = session_key.to_vec();
+
+        // Append a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
+        match pkey.public_params() {
+            PublicParams::X25519 { .. } | PublicParams::X448 { .. } => {}
+            _ => data.extend_from_slice(&checksum::calculate_simple(session_key).to_be_bytes()),
+        }
+
+        let values = pkey.encrypt(rng, &data, EskType::V6)?;
+
+        Ok(PublicKeyEncryptedSessionKey::V6 {
+            packet_version: Default::default(),
+            key_version: Some(pkey.version()),
+            fingerprint: Some(pkey.fingerprint()),
+            pk_algo: pkey.algorithm(),
+            values,
         })
     }
 
+    /// Check if a Key matches with this PKESK's target
+    /// - for v3: is PKESK key id the wildcard, or does it match `id`?
+    /// - for v6: is PKESK fingerprint the wildcard (represented as `None`), or does it match `fp`?
+    pub fn match_identity(&self, key_id: &KeyId, fp: &Fingerprint) -> bool {
+        match self {
+            Self::V3 { id, .. } => id.is_wildcard() || (id == key_id),
+            Self::V6 { fingerprint, .. } => {
+                if let Some(fingerprint) = fingerprint {
+                    fingerprint == fp
+                } else {
+                    true // wildcard always matches
+                }
+            }
+            _ => false,
+        }
+    }
+
     pub fn id(&self) -> &KeyId {
-        &self.id
+        match self {
+            Self::V3 { id, .. } => id,
+            _ => unimplemented!(), // FIXME
+        }
     }
 
-    pub fn mpis(&self) -> &[Mpi] {
-        &self.mpis
+    pub fn fingerprint(&self) -> Option<&Fingerprint> {
+        match self {
+            Self::V6 { fingerprint, .. } => fingerprint.as_ref(),
+            _ => None,
+        }
     }
 
-    pub fn packet_version(&self) -> Version {
-        self.packet_version
+    pub fn values(&self) -> &EskBytes {
+        match self {
+            Self::V3 { values, .. } | Self::V6 { values, .. } => values,
+            _ => unimplemented!(),
```

Comment:
not panic, but return an actual error?

Reply:
:+1: 17d1ba5f346019f8d2f3ce61e199c81b3082b13c